doc/specs/#2563 - closeOnExit and TerminalConnection evolution.md

```diff
@@ -0,0 +1,79 @@
+---
+author: Dustin Howett @DHowett-MSFT
+created on: 2019-07-19
+last updated: 2019-11-05
+issue id: "#2563"
+---
+
+# Improvements to CloseOnExit
+
+## Abstract
+
+This spec describes an improvement to the `closeOnExit` profile feature and the `TerminalConnection` interface that will offer greater flexibility and allow us to provide saner defaults.
+
+## Inspiration
+
+Other terminal emulators like ConEmu have a similar feature.
+
+## Solution Design
+
+1. The existing `closeOnExit` profile key will become an enumerated string key with the following values/behaviours:
+   * `always` - a tab or pane hosting this profile will always be closed when the launched connection terminates.
+   * `graceful` - a tab or pane hosting this profile will be closed **iff** the launched connection closes with an exit code == 0.
+   * `never` - a tab or pane hosting this profile will not automatically close.
+2. The `TerminalConnection` interface will be augmented with an enumerator and a set of events regarding connection state transitions.
+    * enum `TerminalConnection::ConnectionState`
+        * This enum attempts to encompass all potential connection states, even those which do not make sense for a local terminal.
+        * The wide variety of values will be useful to indicate state changes in a user interface.
+        * `NotConnected`: All new connections will start out in this state
+        * `Connecting`: The connection has been initated, but has not yet completed connecting.
+        * `Connected`: The connection is active.
+        * `Closing`: The connection is being closed (usually by request).
+        * `Closed`: The connection has been closed, either by request or from the remote end terminating successfully.
+        * `Failed`: The connection was unexpectedly terminated.
+            * A connection entering the `Failed` state will usually produce a message containing a reason.
+    * event arguments `StateChangedEventArgs(ConnectionState)`
+    * typed event `StateChanged`, whose payload is a `StateChangedEventArgs`
+    * event `TerminalDisconnected` will be removed, as it is replaced by `StateChanged`
+3. As `TerminalApp` is responsible for producing connection instances, _it_ will subscribe to the new connection state event.
+
+The new default value for `closeOnExit` will be `graceful`.
+
+## UI/UX Design
+
+As above. An affordance is provided for the application consuming a connection to determine what state the connection is in and provide sufficient UI. Terminal will not currently make use of any of the new states other than `Closed` and `Failed` to enact UI change.
```

Comment:
> Terminal will not currently make use of any of the new states other than `Closed` and `Failed` to enact UI change.

Do you want to call out in this spec what that UI would be like? As a user, when there's an ungraceful exit, what happens?

When it's set to `never`, and there's a graceful exit, what happens?

Reply:
I've filled out the UI/UX considerations section.